components/ShadowLayer/src/MDCShadowLayer.m

```diff
@@ -304,25 +334,77 @@ - (void)commonLayoutSublayers {
   _topShadow.bounds = bounds;
 
   if (_shadowMaskEnabled) {
-    _bottomShadow.mask = [self shadowLayerMaskForLayer:_bottomShadow];
-    _topShadow.mask = [self shadowLayerMaskForLayer:_topShadow];
+    [self configureShadowLayerMaskForLayer:_topShadowMask];
+    [self configureShadowLayerMaskForLayer:_bottomShadowMask];
   }
   // Enforce shadowPaths because otherwise no shadows can be drawn. If a shadowPath
   // is already set, use that, otherwise fallback to just a regular rect because path.
-  if (!_bottomShadow.shadowPath) {
+  if (!_bottomShadow.shadowPath || _shadowPathIsInvalid) {
     if (self.shadowPath) {
       _bottomShadow.shadowPath = self.shadowPath;
     } else {
       _bottomShadow.shadowPath = [self defaultShadowPath].CGPath;
     }
   }
-  if (!_topShadow.shadowPath) {
+  if (!_topShadow.shadowPath || _shadowPathIsInvalid) {
     if (self.shadowPath) {
       _topShadow.shadowPath = self.shadowPath;
     } else {
       _topShadow.shadowPath = [self defaultShadowPath].CGPath;
     }
   }
+  _shadowPathIsInvalid = NO;
+}
+
+@end
+
+@implementation MDCAnimatedShapeLayer
+
+- (id<CAAction>)actionForKey:(NSString *)key {
+  if ([key isEqualToString:@"path"] || [key isEqualToString:@"shadowPath"]) {
+    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:key];
+    animation.fromValue = [self.presentationLayer valueForKey:key];
+
+    // We have to create a pending animation because if we are inside a UIKit animation block we
+    // won't know any properties of the animation block until it is commited.
+    MDCPendingAnimation *pendingAnim = [[MDCPendingAnimation alloc] init];
+    pendingAnim.animation = animation;
+
+    return pendingAnim;
+  }
+  return [super actionForKey:key];
+}
+
+@end
+
+@implementation MDCPendingAnimation
+
+- (void)runActionForKey:(NSString *)event object:(id)anObject arguments:(NSDictionary *)dict {
+  if ([anObject isKindOfClass:[MDCAnimatedShapeLayer class]]) {
+    MDCAnimatedShapeLayer *layer = (MDCAnimatedShapeLayer *)anObject;
+
+    // In order to synchronize our animation with UIKit animations we have to fetch the resizing
+    // animation created by UIKit and copy the configuration to our custom animation.
+    CALayer *animatedLayer = layer.animationParentLayer;
+    CAAnimation *boundsAction = [animatedLayer animationForKey:@"bounds.size"];
```

Comment:
UIKit will create animations that are additive by default, so if you create your own animation here and attempt to animate it non-additively, there's a decently high likelihood that there are multiple bounds animations associated with the layer if the user tapped quickly.

e.g.:

```
bounds.origin=<CABasicAnimation: 0x60400023a020> bounds.size=<CABasicAnimation: 0x60400003e0e0> bounds.origin-2=<CABasicAnimation: 0x60000022f540> bounds.size-2=<CABasicAnimation: 0x60000022f5c0>
```

I don't know of a clean solution here unfortunately, but it's something to consider.

Reply:
Also this isn't necessarily a new problem in this PR, so I'm not proposing that it need to be fixed here.